### PR TITLE
chore: fix and refactor add generator

### PR DESCRIPTION
**Pull request by rishabh3112**
**What kind of change does this PR introduce?**

refactoring

**Did you add tests for your changes?**
No

**If relevant, did you update the documentation?**
N/A

**Summary**
Updating `prompting()` to use async/await

**Does this PR introduce a breaking change?**

Maybe

**Other information**
#836

### CODE REVIEW

**Comment by rishabh3112** [comment]
Refer #836

**Comment by pranshuchittora** [comment]
@rishabh3112   :+1:

**Comment by rishabh3112** [comment]
Ready for a review now 👍

File: packages/generators/add-generator.ts [hunk 1]
```diff
@@ -116,377 +115,401 @@ export default class AddGenerator extends Generator {
...
+				} = await this.prompt([
+					Input("topScope", "What do you want to add to topScope?"),
+				]);
+
+				this.configuration.config.topScope.push(topScopeAnswer.topScope);
+				return;
+			}
+		}
+
+		const temp: string = action;
```
Comment (evenstensberg) [inline packages/generators/add-generator.ts]: reassigning a const

Comment (rishabh3112) [reply packages/generators/add-generator.ts]: Where? temp is never reassigned anywhere and here it is declared.

**Comment by rishabh3112** [comment]
Thanks for review @ematipico 
I will look into it and resolve these in this PR. 

P.S
Just for clearing things 😅,
All of the code, in which you requested changes, was there as it is before this PR and not written by me originally.

**Comment by ematipico** [comment]
Oh yeah don't worry! We know it's working all right, if you think it doesn't require any change and it will still too much time of yours,  mark it as resolved and move along. :)

File: packages/generators/add-generator.ts [hunk 2]
```diff
@@ -116,377 +115,401 @@ export default class AddGenerator extends Generator {
...
+				defOrPropDescription = Object.keys(defOrPropDescription[0].enum)
+					.map((p: string): object => {
+						return Object.assign(
+							{},
+							{
+								[originalPropDesc[p]]: "noop",
+							},
+						);
+					})
+					.reduce((result: object, currentObject: object): object => {
```
[the remark this answers is not in the document]
Comment (rishabh3112) [reply packages/generators/add-generator.ts]: Well here we are reducing the objects in the array to single object with all the properties.

So, this requires two loops:
1. for iterating over the array of objects.
2. to add the props of that object to final object. Here we can also use `Object.assign` but that also takes `O(m)` complexity internally, where m is number of properties.

So minimum complexity that I can think of is `O(nm)` which is also of the current code.
where n = number of objects in array
          m = maximum number of properties that a object has in this array

Is there any other algorithm for this task in your mind @ematipico ?

**Comment by rishabh3112** [comment]
Well all are done and it didn't steal my time :). Pushing commits after a quick test.

**Comment by ematipico** [comment]
We should make a separate PR to address these security issues IF there's already some patch for packages involved

**Comment by rishabh3112** [comment]
@ematipico Synk wasn't able to process the package.json. I can't see if this error due to a security flaw in any package. Here is the error description:
image

**Comment by ematipico** [comment]
Does your `npm audti` shows some vulnerabilities? If so, that's the reason

**Comment by rishabh3112** [comment]
Thanks, have done it @dhruvdutt, Is there any more changes required?

File: packages/generators/add-generator.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rishabh3112) [reply packages/generators/add-generator.ts]: @ematipico?

[the remark this answers is not in the document]
Comment (ematipico) [reply packages/generators/add-generator.ts]: Thanks for explaining it. I guess we can live with it. It's not like we will have tons of properties. No need to optimise at the moment

**Comment by ematipico** [comment]
Wanna rebase? CI is fixed

**Comment by rishabh3112** [comment]
I will do it then :)

**Comment by rishabh3112** [comment]
Done @ematipico

**Comment by rishabh3112** [comment]
Will do it again then :+1:

**Comment by rishabh3112** [comment]
Done @evenstensberg

**Comment by rishabh3112** [comment]
@webpack/cli-team can I have a review on this PR, this is fine IMO

**Comment by rishabh3112** [comment]
cc @evenstensberg

**Comment by rishabh3112** [comment]
@evenstensberg PTAL 😅

**Comment by evenstensberg** [comment]
Thanks, will get back to you this week. Quite a big PR and need to manually test

**Comment by evenstensberg** [comment]
@rishabh3112 could you rebase and let's continue on getting this merged once you're done?

**Comment by rishabh3112** [comment]
Sure, will catch you tommorow. It's night here now and I am sleeping :)

**Comment by evenstensberg** [comment]
Gotcha, sleep tight!

File: tsconfig.json [hunk 3]
```diff
@@ -1,6 +1,7 @@
 {
 	"extends": "./tsconfig.base.json",
 	"compilerOptions": {
-		"allowJs": true
+		"allowJs": true,
+		"resolveJsonModule": true
```
Comment (evenstensberg) [inline tsconfig.json]: This is already added in the master branch

Comment (rishabh3112) [reply tsconfig.json]: What? I can't see `resolveJsonModule` on master. Have merged with master only!

Comment (evenstensberg) [reply tsconfig.json]: `tsconfig.base.js`

Comment (rishabh3112) [reply tsconfig.json]: Okay, will remove :)

Comment (evenstensberg) [reply tsconfig.json]: <3

**Comment by evenstensberg** [comment]
Could you rebase?

**Comment by rishabh3112** [comment]
Will do it @evenstensberg, but is it required now?, as add package is being removed Or we would keep generators and remove packages only?

**Comment by evenstensberg** [comment]
Nah I don't think it's worth

**Pull request closed without merging**